### DISCUSSION TITLE
Expiration Date Reminders for Documents and Licenses

### DISCUSSION THREAD

**Discussion by the-kizz** [bitwarden/clients#5983] [Password Manager]
### Code Contribution Proposal

I am proposing the introduction of an "expiry date" feature for identity-related items such as licenses. This feature, when integrated, will enable apps to dispatch push notifications or reminders as the expiration dates of these documents approach.

this is some of my thoughts:

1. **Client-Side Implementation:** This is the most feasible approach given the limitations around encryption. By running the logic on the client side, the data doesn't need to be decrypted by the server.

      **Field for Expiry Date:** a custom field named "Expiry" could be created by users. This would store the expiry date of documents or licenses. or it could be a permanent optional field.

      **Client-Side Script:** Develop a script or add a feature within the Bitwarden client application (be it web, desktop, or mobile). This script or feature would:

      - Fetch Items: Retrieve all items from the local vault.
      - Parse Items: Look for the field "Expiry" within each item.
      - Compare Dates: Compare the expiry date with the current date.
      - Send Alert: If the date is nearing expiration, the client application could send a local push notification or display an alert to the user.

     **Color-Coding:** As a visual cue, the client application could color-coded items nearing expiration, with colours growing more intense as the expiry date approaches.

2. **SMTP Alerts:** The suggestion to use SMTP for email alerts is feasible, but it'd require decryption of the expiry date. Given that this can't be done server-side, it might be possible on the client side, though it would introduce challenges around integrating and securing SMTP functionalities directly within the client apps.

3. **Web UI Implementation:** The feature could be implemented in the Bitwarden web UI. This would involve:

      **Fetch and Parse:** When users access their vault via the web, the UI could fetch and parse the expiry field.

      **Display:** Show items that are approaching their expiry dates, possibly highlighting them with different colours with a traffic light system.

**Comment by dwbit** [comment]
Hey @conroyke56 are you proposing an idea for code that you would like to contribute, or voicing a feature request? 

The Bitwarden Github discussions space is used only for code contributions that you wish to write. For feature requests, please visit: https://bitwarden.com/featurerequest

**Reply by the-kizz** [reply]
Sorry. Saw a lot of proposal tags. Thought needed to propose a change first. 

If it’s of interest to people. I’d attempt to write the code. Yeh.

**Comment by the-kizz** [comment]
**Implementation for Expiry Notifications with User Consent**:

1. **Model Updates**:

For `Identity`:
```rust
pub expiry_date_encrypted: Option<CipherString>,
pub expiry_date_plaintext: Option<String>,  // Populated only if notifications are opted in.
pub expiry_status: u8,
pub plaintext_title: Option<String>,  // Populated only if notifications are opted in.
```

For `IdentityView`:
```rust
pub expiry_date: Option<String>,  // Used for decrypted view and data input.
pub expiry_status: u8,
```

2. **Logic for Setting Expiry**:
- Any user can set the expiry date using the `IdentityView`'s `expiry_date` field.
- This date is always encrypted and stored in `Identity`'s `expiry_date_encrypted`.

3. **Logic to Populate `expiry_date_plaintext` and `plaintext_title`**:
- If a user wants to enable expiry notifications for a document:
  1. Present a warning: "Enabling expiry notifications will store the expiry date and the title in plain text. Are you sure you want to continue?"
  2. If the user agrees:
     - Populate `Identity`'s `expiry_date_plaintext` from `IdentityView`'s `expiry_date`.
     - Also populate `Identity`'s `plaintext_title` from the document's title.
  3. If the user disagrees, notifications won't be enabled, and plain text fields remain unpopulated.

4. **Expiry Status Utilization**:
- The `expiry_status` field serves as a grade system for notifications:
  - `1`: Valid (more than three weeks to expiry)
  - `2`: Three weeks to expiry
  - `3`: One week to expiry
  - `4`: Two days to expiry
  - `5`: Expired today
  - `6`: Expired
- A daily background task checks `expiry_date_plaintext` against the current date and updates the `expiry_status` accordingly. 

5. **Notification Logic**:
- If `expiry_date_plaintext` is populated (indicating notifications are enabled):
  - Use it along with `expiry_status` and `plaintext_title` to send graded expiry reminders based on the aforementioned `expiry_status` values.

This approach ensures that only users who explicitly opt-in for expiry notifications will have their expiry date and title stored in plain text.

---
Once that’s implemented, I’d need to start working on the front end. 

This logic could be implemented globally across the whole system, but I don’t think I have the capabilities to code that.

**Reply by dwbit** [reply]
Thanks for clarification! I've added the 'proposed' tag and the team will get back to you shortly.

**Reply by vadimkholodilo** [reply]
One thing which I would like to express is that, please, give any other way to see entries which are about to expire in addition to color-code them. It'll help a lot people with color blindness and those who use screen readers

**Reply by the-kizz** [reply]
Good point. I didn’t consider that. Would be easy enough to add a days remaining calculation.

**Comment by dwbit** [comment]
Closing out this older request, feel free to ping if you need anything.
